src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -0,0 +1,197 @@
+/// 4  Compute bit permutation that sorts i..i+`num_multi_bits`
+/// 5. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, n-1th composition is returned. This is the permutation which sorts the inputs
+///
+/// # Errors
+/// If any underlying protocol fails
+/// # Panics
+/// Panics if input doesn't have same number of bits as `num_bits`
+
+pub async fn generate_permutation_opt<F>(
+    ctx: SemiHonestContext<'_, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    F: Field,
+{
+    let ctx_0 = ctx.narrow(&Sort(0));
+    assert_eq!(sort_keys.len(), num_bits as usize);
+
+    let last_bit_num = if num_multi_bits > num_bits {
+        num_bits
+    } else {
+        num_multi_bits
+    };
+
+    let lsb_permutation = multi_bit_permutation(
+        ctx_0.narrow(&BitPermutationStep),
+        &sort_keys[0..last_bit_num.try_into().unwrap()],
+    )
+    .await?;
+
+    let input_len = u32::try_from(sort_keys[0].len()).unwrap(); // safe, we don't sort more that 1B rows
+
+    let mut composed_less_significant_bits_permutation = lsb_permutation;
+    for bit_num in (num_multi_bits..num_bits).step_by(num_multi_bits.try_into().unwrap()) {
+        let ctx_bit = ctx.narrow(&Sort(bit_num));
+        let revealed_and_random_permutations = shuffle_and_reveal_permutation(
+            ctx_bit.narrow(&ShuffleRevealPermutation),
+            input_len,
+            composed_less_significant_bits_permutation,
+        )
+        .await?;
+
+        let (randoms_for_shuffle0, randoms_for_shuffle1, revealed) = (
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .0
+                .as_slice(),
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .1
+                .as_slice(),
+            revealed_and_random_permutations.revealed.as_slice(),
+        );
+
+        let last_bit_num = if bit_num + num_multi_bits > num_bits {
+            num_bits
+        } else {
+            bit_num + num_multi_bits
+        };
+
+        let futures =
+            (bit_num..last_bit_num)
+                .zip(repeat(ctx_bit.clone()))
+                .map(|(idx, ctx_bit)| async move {
+                    secureapplyinv(
+                        ctx_bit.narrow(&MultiApplyInv(idx)),
+                        sort_keys[idx as usize].clone(),
+                        (randoms_for_shuffle0, randoms_for_shuffle1),
+                        revealed,
+                    )
+                    .await
+                });
+        let bits_i_sorted_by_less_significant_bits = try_join_all(futures).await?;
+
+        let bits_i_permutation = multi_bit_permutation(
+            ctx_bit.narrow(&BitPermutationStep),
+            &bits_i_sorted_by_less_significant_bits,
+        )
+        .await?;
+
+        let composed_i_permutation = compose(
```

Comment:
These names which include an `_i_` don't make sense now that we are doing multi-bit. I'd suggest:

`next_few_bits_sorted_by_less_significant_bits`

and

`next_few_bits_permutation`